public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,493 @@
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+    /* using absolute positioning because flex interferes with ellipsis */
+    position: absolute;
+    width: 50%;
+    right: 0;
+    text-align: right;
+    transition: opacity 100ms linear;
+    opacity: 0;
+  `,
+  statusShowing: css`
+    opacity: 1;
+  `,
+  error: css`
+    color: ${theme.palette.brandDanger};
+  `,
+  valueCell: css`
+    overflow: hidden;
+    text-overflow: ellipsis;
+  `,
+  valueList: css`
+    margin-right: ${theme.spacing.sm};
+  `,
+  valueListWrapper: css`
+    padding: ${theme.spacing.sm};
+    & + & {
+      border-left: 1px solid ${theme.colors.border2};
+    }
+  `,
+  valueListArea: css`
+    display: flex;
+    flex-direction: row;
+    margin-top: ${theme.spacing.sm};
+    overflow-x: auto;
+  `,
+  valueTitle: css`
+    margin-left: -${theme.spacing.xs};
+    margin-bottom: ${theme.spacing.sm};
+  `,
+  validationStatus: css`
+    padding: ${theme.spacing.xs};
+    margin-bottom: ${theme.spacing.sm};
+    color: ${theme.colors.textStrong};
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  `,
+}));
+
+export class UnthemedLokiLabelBrowser extends React.Component<BrowserProps, BrowserState> {
+  state = {
+    labels: [] as SelectableLabel[],
+    searchTerm: '',
+    status: 'Ready',
+    error: '',
+    validationStatus: '',
+  };
+
+  onChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
+    this.setState({ searchTerm: event.target.value });
+  };
+
+  onClickRunLogsQuery = () => {
+    const selector = buildSelector(this.state.labels);
+    this.props.onChange(selector);
+  };
+
+  onClickRunMetricsQuery = () => {
+    const selector = buildSelector(this.state.labels);
+    const query = `rate(${selector}[$__interval])`;
+    this.props.onChange(query);
+  };
+
+  onClickClear = () => {
+    this.setState((state) => {
+      const labels: SelectableLabel[] = state.labels.map((label) => ({
+        ...label,
+        values: undefined,
+        selected: false,
+        loading: false,
+        hidden: false,
+        facets: undefined,
+      }));
+      return { labels, searchTerm: '', status: '', error: '', validationStatus: '' };
+    });
+    store.delete(LAST_USED_LABELS_KEY);
+  };
+
+  onClickLabel = (name: string, value: string | undefined, event: React.MouseEvent<HTMLElement>) => {
+    // Stop early before it reaches the ClickOutsideWrapper
+    event.stopPropagation();
+    const label = this.state.labels.find((l) => l.name === name);
+    if (!label) {
+      return;
+    }
+    // Toggle selected state
+    const selected = !label.selected;
+    let nextValue: Partial<SelectableLabel> = { selected };
+    if (label.values && !selected) {
+      // Deselect all values if label was deselected
+      const values = label.values.map((value) => ({ ...value, selected: false }));
+      nextValue = { ...nextValue, facets: 0, values };
+    }
```

Comment:
I think if we created a helper function in `utils` to hide all of this logic, the readability would be improved.